R/util_data_summaries.R

```diff
@@ -126,3 +155,509 @@ ds.summarize_by_column <- function(
+                )
+        }
+    } else{
+        msl$int_apa <- ""
+    }
+    
+    return(msl)
+}
+
+#########################################################
+###
+###     General Model Summary Function(s)
+###     Should work on models outside of glm1 spec.
+###
+#########################################################
+
+ds.model_summary_df <- function( 
+    models, # list of models
+    summary_func = ds.summarize_glm1  # extracts ds.glm1 features
+){
+    # returns a data.frame with one row per supplied model
+    # and one column for each list index returned by the
+    # supplied summary_func
+    # example use cases of output:
+    #   sort or filter summary data.frame for specific models
+    #   pass to printing functions to nice summary tables
+    
+    # msls = model summary lists
+    msls <- lapply( models , function(glm_model){
+        as.data.frame(summary_func(glm_model))
+    })
+    
+    # remove empty feature lists
+    # (e.g., may happen if ds.glm1 spec is violated)
+    sanitized_msls <- list()
+    for( i in 1:length(msls) ){
+        if( nrow(msls[[i]]) > 0 ){
+            sanitized_msls[[length(sanitized_msls) + 1]] <- msls[[i]]
+        }
+    }
+    if(length(sanitized_msls) > 1){
+        feature_table <- util.rbind_union(sanitized_msls)
+    } else if(length(sanitized_msls) == 1){
+        feature_table <- sanitized_msls[[1]]
+    } else{
+        feature_table <- data.frame()
+    }
+    return(feature_table)
+}
+
+
+#########################################################
+###
+###     Unit Tests
+###
+#########################################################
+
+ds.helper$unit_test <- function(){
+    
+    assert_expected_output <- function( func_name , input, expected_output ){
+        # warn if evaluated func_name(input) != expected_output
+        # errors messages are not compared, but errors can be expected
+        
+        # func_name passed as string rathern than func itself
+        # so the func_name can be included warnings
+        func <- eval(parse(text=func_name))
+        
+        real_output <- tryCatch(
+                            func(input),
+                            error = function(e) e
+                        )
+        
+        both_are_errors <-  any(class(real_output) %in% "error") & 
+                            any(class(expected_output) %in% "error")
+        outputs_identical <- identical(expected_output, real_output)
+        
+        error_message <- ""
+        if("error" %in% class(real_output)){
+            error_message <- "\nError raised in tested function: " %+%
+                real_output$message
+        }
+        
+        if( ! ( both_are_errors | outputs_identical ) ){
+            "assert_expected_output failed in " %+%
+                func_name %+% " when passed '" %+%
+                paste(input,collapse=", ") %+% "'" %+% 
+                ". \n" %+% error_message %>%
+                util.warn()
+        }
+    }
```

Comment:
I'm sure there are good arguments to be made for using `Runit`. Here are two possible reasons _not_ to refactor for Runit:

1. It's yet another package that `gymanst` would depend on. Generally speaking, more dependencies = bad.
    1. Packages sometimes get stale and break in new versions. If a package is popular and well-maintained and/or it does something really complicated that would take a long time to code, then we should obviously use it. (Aside: [RUnit is downloaded 8,000/month but is "no longer actively developed"](https://cran.r-project.org/web/packages/RUnit/README.html)). Regardless of the specific case here, I just want to mention for posterity (and I'm guessing you probably agree with me already) that our "default" should _not_ be to use an external package to do something that we can easily do ourselves. Personally, I think there are few worse things in coding than running code that's been working great for years and having it break because somebody else messed up a dependency. Then, instead of getting to do the analysis work you expected to (and are probably are on a deadline for), you have to dig through the bowels of your code and refactor it because somebody else, far away, messed up. 
    2. More packages mean slower load times and higher risk for name-space conflicts.
2. This is already written and appears to work. Rewriting it to use a different approach would take more time investment.


If you think there are big advantages to using RUnit, you're probably right. Maybe you could lead a workshop on it, and we could all together hash out pros and cons. (Please just make sure I can come to it because that's a conversation I wouldn't want to miss.)